### PR TITLE
MDMM tests and hardware-aware metric functions

### PR DESCRIPTION
**Pull request by ArghyaRanjanDas**
Adds tests for the existing MDMM pruning method and integrates the hardware-aware metric functions (`FPGAAwareSparsityMetric` and `PACAPatternMetric`) that were lost during intermediate branch merging.                                             
  
Also fixes a `float32/int32` dtype bug in `get_layer_sparsity` on the TensorFlow backend.

  ### Supported backends
  - `KERAS_BACKEND=tensorflow`                                
  - `KERAS_BACKEND=torch`                                 
                                                              
  ### Test files
  - `pytest tests/test_mdmm.py tests/test_mdmm_metrics.py`

### CODE REVIEW

File: src/pquant/pruning_methods/metric_functions.py [hunk 1]
```diff
@@ -83,3 +86,145 @@ def __call__(self, weight):
...
+
+    Models how weights are packed into DSP blocks (groups of size `rf`) and
+    further into BRAM blocks (groups of `c` DSP blocks, where `c` derives from
+    `bram_width` and `precision`). Returns the fraction of zero-valued groups
+    at the chosen target_resource level.
+    """
+
+    def __init__(self, rf=1, precision=16, target_resource='DSP',
+                 bram_width=36, epsilon=1e-3):
+        assert target_resource in ['DSP', 'BRAM'], "target_resource must be 'DSP' or 'BRAM'."
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/metric_functions.py:102]: Seems to be overkill to assert additional parameters asserts of FPGAAwareSparsityMetric, taking into account that Pydantic classes are used for this ( to validate after config upload ).

I would recommend to set numerical values in the model with additional parameters _ge_ for > , _le_ for <. Also, the _target_resource_ value must be already validated, as you are using Literal.

File: src/pquant/pruning_methods/utils/patterns.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+    if perm == (0, 1, 2, 3):                   # identity permutation
+        return w
+    return ops.transpose(w, perm)
+
+
+def _get_kernels_and_patterns(w, src="OIHW", epsilon=1e-5):
+    # src:
+    #   PyTorch: (out, in, kH, kW): OIHW
+    #   Keras  : (kH, kW, in, out): HWIO
+    w_permuted = convert_conv_layout(w, src="OIHW", dst="OIHW")
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/utils/patterns.py:39]: In this command line,  src="OIHW" and dst="OIHW", so basically there is no permutation conducted. Do we need this?

File: src/pquant/pruning_methods/utils/patterns.py [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+    C_out, C_in, kH, kW = ops.shape(w_permuted)
+    kernels = ops.reshape(w_permuted, (C_out * C_in, -1))
+    all_patterns = ops.cast(ops.greater(ops.abs(kernels), epsilon), dtype="uint8")
+
+    return kernels, all_patterns, (C_out, C_in, kH, kW)
+
+def _get_unique_patterns_with_counts(all_patterns):
+    """Returns the unique patterns and their counts."""
+    np_patterns = ops.convert_to_numpy(all_patterns)
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/utils/patterns.py:48]: Can be problematic if we want to track a computational graph

File: src/pquant/pruning_methods/utils/patterns.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    # Cast uint8 patterns/kernel-binarization to the kernel float dtype so
+    # arithmetic with the float kernels works under TF's strict-dtype rules.
+    Tk = ops.cast(Tk, k.dtype)
+    P = ops.cast(P, k.dtype)
+    Tk_exp = ops.expand_dims(Tk, 1)
+    k_exp = ops.expand_dims(k, 1)
+    P_exp = ops.expand_dims(P, 0)
+
+
+    if distance_metric == 'hamming':
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/utils/patterns.py:114]: I would add distance_metic dict in constants and reuse it here in the code

File: src/pquant/pruning_methods/metric_functions.py [hunk 5]
```diff
@@ -83,3 +86,145 @@ def __call__(self, weight):
...
+        padding_needed = (self.rf - num_weights % self.rf) % self.rf
+        weight_padded = ops.pad(weight_reshaped, [[0, 0], [0, padding_needed]])
+
+        return weight_padded
+
+    def __call__(self, weight):
+        prepared_weights = self._prepare_weights(weight)
+        dsp_groups = ops.reshape(prepared_weights, (prepared_weights.shape[0], -1, self.rf))
+
+        if self.target_resource == 'DSP':
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/metric_functions.py:146]: the same here, target_resource dict in constants and reuse in the code,  strings aren't prefered

File: src/pquant/pruning_methods/metric_functions.py [hunk 6]
```diff
@@ -83,3 +86,145 @@ def __call__(self, weight):
...
+
+    Selects a small set of dominant binary patterns over kernel layouts on the
+    first call (cached for the metric instance lifetime). Returns the mean
+    distance of every kernel to its closest dominant pattern. Operates on 4D
+    Conv2D weights only; returns 0 for non-4D inputs.
+    """
+
+    def __init__(self, num_patterns_to_keep=16, beta=0.75,
+                 epsilon=1e-5, distance_metric='valued_hamming'):
+        assert num_patterns_to_keep > 0, "num_patterns_to_keep must be > 0"
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/metric_functions.py:183]: same situation as in FPGAMetic, this evaluation should happen on the level of Pydantic class

File: src/pquant/pruning_methods/mdmm.py [hunk 7]
```diff
@@ -44,6 +45,14 @@ def build(self, input_shape):
             "l0_mode": l0_mode,
             "scale_mode": scale_mode,
             "rf": pruning_parameters.rf,
+            # FPGAAwareSparsityMetric
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/mdmm.py:48]: I would rather use sth like metric_type parameter that basically is an object of the class ( _PACAPatternMetric_ or _FPGAAwareSparsityMetric_ ) , and then use corresponding parameters from it within the code. For example, metric_type.num_patterns_to_keep, etc. And initialization of parameters for these classes can happen automatically during load of the config file, for the reference how it's done in _PQConfig.load_from_file method._

File: src/pquant/pruning_methods/mdmm.py [hunk 8]
```diff
@@ -54,6 +63,11 @@ def build(self, input_shape):
         else:
             raise ValueError(f"Unknown metric_type: {metric_type}")
 
+        # PACA always pairs with EqualityConstraint at target 0 (preserves original design).
+        if metric_type == "PACAPatternSparsity":
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/mdmm.py:67]: why not to use this name from the metrics_dict in the constants and  take _constraint_type_ and _target_value_ from class parameters?

File: src/pquant/pruning_methods/metric_functions.py (hunk 6 above)
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/metric_functions.py:183]: back to previous comments, all validation should be done in Pydantic class

File: src/pquant/configs/config_mdmm_paca.yaml [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+  rf: 1
+  epsilon: 1.0e-03
+  scale: 1.0
+  damping: 1.0
+  use_grad: false
+  constraint_lr: 1.0e-3
+  # PACAPatternMetric-specific
+  num_patterns_to_keep: 16
+  beta: 0.85
+  distance_metric: "cosine"   # "hamming", "valued_hamming", or "cosine"
```
Comment (nastiapetrovych) [inline src/pquant/configs/config_mdmm_paca.yaml:21]: these details should be also added on the documentation page

File: src/pquant/data_models/pruning_model.py [hunk 10]
```diff
@@ -89,3 +91,11 @@ class MDMMPruningModel(BasePruningModel):
     l0_mode: Literal["coarse", "smooth"] = Field(default="coarse")
     scale_mode: Literal["mean", "sum"] = Field(default="mean")
     constraint_lr: float = Field(default=1.0e-3)
+    # FPGAAwareSparsityMetric (only used when metric_type == "FPGAAwareSparsity")
+    precision: Optional[int] = Field(default=None)
```
Comment (nastiapetrovych) [inline src/pquant/data_models/pruning_model.py:95]: I would add here metric_type and move these parameters to corresponding metric classes: FPGAAwareSparsityMetric and PACAPatternMetric

File: src/pquant/pruning_methods/utils/patterns.py [hunk 11]
```diff
@@ -0,0 +1,155 @@
+from functools import lru_cache
+import numpy as np
+import keras
+from keras import ops
+
+_AX = {"H": 0,   # kernel height  (kH)
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/utils/patterns.py:6]: can be moved to constants?

File: src/pquant/pruning_methods/metric_functions.py [hunk 12]
```diff
@@ -83,3 +86,145 @@ def __call__(self, weight):
...
+        dsp_groups_padded = ops.pad(dsp_groups, [[0, 0], [0, bram_padding], [0, 0]])
+        bram_groups = ops.reshape(dsp_groups_padded, (dsp_groups.shape[0], -1, self.c, self.rf))
+
+        bram_group_norms = ops.sqrt(ops.sum(ops.square(bram_groups), axis=(-1, -2)))
+        zero_bram_groups = ops.less_equal(bram_group_norms, self.epsilon)
+        num_bram_groups = ops.cast(ops.size(bram_group_norms), dsp_groups.dtype)
+        return ops.sum(ops.cast(zero_bram_groups, dsp_groups.dtype)) / num_bram_groups
+
+
+class PACAPatternMetric:
```
Comment (nastiapetrovych) [inline src/pquant/pruning_methods/metric_functions.py:172]: i would say so make these metric classes pydantic

**Comment by nastiapetrovych** [comment]
We can't merge directly to main branch unless it was merged to a dev one @ArghyaRanjanDas

**Comment by nastiapetrovych** [comment]
After testing MDMM tests and training within model training loop, I detected an issue with _PydanticSerializationUnexpectedValue_,  especially:
1) `Expected `ActivationPruningModel` - serialized value may not be as expected [field_name='pruning_parameters', input_value=MDMMPruningModel(disable_...e, distance_metric=None), input_type=MDMMPruningModel]).`
2) `PydanticSerializationUnexpectedValue(Expected `enum` - serialized value may not be as expected [field_name='constraint_type', input_value='Equality', input_type=str])`

**Review by nastiapetrovych — CHANGES_REQUESTED** [review]
Hi @ArghyaRanjanDas ,
There are several several modifications that should be corrected before merging to the dev branch.

**Comment by nastiapetrovych** [comment]
This PR #37  was closed as @ArghyaRanjanDas opened another one -> #46 from the updated dev branch.
